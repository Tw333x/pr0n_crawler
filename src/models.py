import os

from peewee import *
from playhouse.fields import ManyToManyField
from playhouse.sqlite_ext import SqliteExtDatabase

__all__ = [
    'db',
    'Site',
    'Video',
    'Tag',
    'VideoToTag'
]

db = SqliteExtDatabase(
    os.path.join(os.path.dirname('..'), 'database.sqlite')
)


class BaseModel(Model):
    class Meta:
        database = db


class Site(BaseModel):
    name = CharField()
    url = CharField()


class Tag(BaseModel):
    tag = CharField()
    slug = CharField()


class Video(BaseModel):
    title = CharField()
    duration = IntegerField()
    url = CharField()
    thumbnail_url = CharField()

    site = ForeignKeyField(Site)
    tags = ManyToManyField(Tag, related_name="videos")


VideoToTag = Tag.videos.get_through_model()
