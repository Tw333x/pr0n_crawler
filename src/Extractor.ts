import { Site } from './Site';

class Extractor {
  constructor(private site: Site, private content: string) {
  }

  public extractPreviousPage() {
    const { selector, normalizer } = this.site.getFields().previousPage;

  }

  public extractVideosUrl() {
    const { selector, normalizer } = this.site.getFields().previousPage;

  }

  public extractVideosThumbnailUrl() {
    const { selector, normalizer } = this.site.getFields().previousPage;

  }
}

export {
  Extractor,
};
