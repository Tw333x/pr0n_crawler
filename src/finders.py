def find_videos_title(tree, video_title_selector):
    """
    :type tree: lxml.html.Element
    :type video_title_selector: str
    :rtype: list[str]
    """

    return tree.xpath(video_title_selector)


def find_videos_duration(tree, video_duration_selector):
    """
    :type tree: lxml.html.Element
    :type video_duration_selector: str
    :rtype: list[str]
    """

    return tree.xpath(video_duration_selector)


def find_videos_url(tree, video_url_selector):
    """
    :type tree: lxml.html.Element
    :type video_url_selector: str
    :rtype: list[str]
    """

    return tree.xpath(video_url_selector)


def find_videos_thumbnail_url(tree, video_thumbnail_url_selector):
    """
    :type tree: lxml.html.Element
    :type video_thumbnail_url_selector: str
    :rtype: list[str]
    """

    return tree.xpath(video_thumbnail_url_selector)


def find_video_details(tree, selectors):
    """
    :type tree: lxml.html.Element
    :type selectors: dict[str, str]
    :rtype: dict[str, list]
    """

    return dict(
        tags=tree.xpath(selectors.get('video_details_tags'))
    )


def find_prev_page(tree, prev_page_selector):
    """
    :type tree: lxml.html.Element
    :type prev_page_selector: str
    :rtype: str
    """

    return tree.xpath(prev_page_selector)[0]
