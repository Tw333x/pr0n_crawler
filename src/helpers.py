def hms_to_s(hms: str) -> int:
    """
    Convert hh:mm:ss to sum of seconds
    :param hms:
    :return:
    """
    l = list(map(int, hms.split(':')))
    return sum(n * sec for n, sec in zip(l[::-1], (1, 60, 3600)))
