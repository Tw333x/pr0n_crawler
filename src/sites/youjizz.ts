import { Site } from '../Site';

class YouJizzSite extends Site {
  public getUrl(): string {
    return 'https://www.youjizz.com';
  }

  public getName(): string {
    return 'YouJizz';
  }

  public getFavicon(): string {
    return 'https://www.youjizz.com/favicon.ico';
  }

  public getEntryPoint(): string {
    return 'https://www.youjizz.com/newest-clips/100.html';
  }

  public getFields(): SiteFields {
    return {
      previousPage: {
        selector: '(//ul[contains(@class, "pagination")]/li[contains(@class, "active")]/a/@href)[last()]',
        normalizer: (previousPage: string) => String(new URL(previousPage, this.getUrl())),
      },
      videosUrl: {
        selector: '//*[contains(@class, "video-title")]/a/@href',
        normalizer: (urls: Array<string>) => urls.map(url => String(new URL(url, this.getUrl()))),
      },
      videosThumbnailUrl: {
        selector: '',
        normalizer: (thumbnailsUrl: Array<string>) => thumbnailsUrl,
      },
      video: {
        title: {
          selector: '//*[contains(@class, "video-player")]/following-sibling::h3/text()',
          normalizer: (title: string) => title,
        },
        duration: {
          selector: '',
          normalizer: (duration: string) => Number(duration),
        },
        tags: {
          selector: '',
          normalizer: (tags: Array<string>) => tags,
        },
      },
    };
  }
}

export { YouJizzSite };
