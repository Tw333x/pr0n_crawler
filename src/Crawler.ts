import syncRequest from 'sync-request';
import { zip } from 'zip-array';
import { apiKey, graphqlEndpoint } from '../config';
import { Site as SiteEntity } from './entities/Site';
import { Video } from './entities/Video';
import ExtractorNoPreviousPageFoundError from './errors/ExtractorNoPreviousPageFoundError';
import { Extractor } from './Extractor';
import { Site } from './Site';

class Crawler {
  private crawledPages: number;
  private crawledVideos: number;

  constructor(private site: Site) {
    this.crawledPages = 0;
    this.crawledVideos = 0;
  }

  public crawl(url: string = this.site.getEntryPoint()) {
    console.log(`Fetching ${url}...`);

    const extractor = new Extractor(this.site, syncRequest('GET', url).getBody('utf8'));
    const videos: Video[] = this.initVideos(extractor);

    videos.map(async video => {
      const response = syncRequest('GET', video.url);
      this.handleVideo(video, response.getBody('utf8'));
      this.crawledVideos += 1;
    });

    this.crawledPages += 1;
    console.log(`Fetching ${url}: done.`);

    try {
      const previousPage = extractor.extractPreviousPage();
      this.crawl(previousPage);
    } catch (e) {
      if (e instanceof ExtractorNoPreviousPageFoundError) {
        console.info(`Crawling ${this.site.getName()} done.`);
      } else {
        throw e;
      }
    }
  }

  private initVideos(extractor: Extractor) {
    const siteEntity = new SiteEntity();

    siteEntity.url = this.site.getUrl();
    siteEntity.name = this.site.getName();
    siteEntity.favicon = this.site.getFavicon();

    return zip(extractor.extractVideosUrl(), extractor.extractVideosThumbnailUrl()).map(payload => {
      const video = new Video();

      video.site = siteEntity;
      video.url = payload[0];
      video.thumbnailUrl = payload[1];

      return video;
    });
  }

  private handleVideo(video: Video, content: string) {
    const videoExtractor = new Extractor(this.site, content);

    video.title = videoExtractor.extractVideoTitle();
    video.tags = videoExtractor.extractVideoTags();
    video.duration = videoExtractor.extractVideoDuration();

    const mutation = `
      mutation CreateVideo($input: CreateVideoInput!) {
        createVideo(input: $input) {
          id title url thumbnailUrl duration
          site { id name host }
          tags { id tag slug }
        }
      }
    `;

    const variables = {
      input: {
        title: video.title,
        url: video.url,
        thumbnailUrl: video.thumbnailUrl,
        duration: video.duration,
        site: {
          name: video.site.name,
          host: video.site.url,
        },
        tags: video.tags,
      },
    };

    console.log(`Sending video ${video.title} to GraphQL API...`);

    const result = syncRequest('POST', graphqlEndpoint, {
      headers: {
        'x-auth-token': apiKey,
      },
      json: {
        variables,
        query: mutation,
      },
    });

    try {
      const body = result.getBody('utf8');
      console.info(`Sending video ${video.title} to GraphQL API: done`);
      console.info(body);
    } catch (e) {
      console.error(`Sending video ${video.title} to GraphQL API: error`);
      console.error(e);
      process.exit(1);
    }
  }
}

export { Crawler };
