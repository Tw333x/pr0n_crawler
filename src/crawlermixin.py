import asyncio
import logging
import time

import aiohttp
from inflection import humanize, parameterize
from lxml import html
from tenacity import before_log, retry, stop_after_attempt, wait_random

from src.models import Video, Site, Tag, VideoTag


class CrawlerMixin(object):
    site_name = None  # type: str
    site_url = None  # type: str
    crawler_entry_point = None  # type: str
    crawler_selectors = dict()  # type: dict[str, str | dict[str, str]]
    crawler_max_videos = 9000

    already_existing_videos_count = 0

    def __init__(self):
        self.crawler_current_videos = 0
        self._hydrate_logger()

        if not (self.site_name or self.site_url):
            raise ValueError("Site's name and site's url should not be None.")

        self.site, created = Site.get_or_create(
            name=self.site_name,
            url=self.site_url
        )

        if created:
            self.logger.info('Site created.')

    @retry(
        stop=stop_after_attempt(20), wait=wait_random(8, 512),
        before=before_log(logging.getLogger(), logging.WARN)
    )
    async def crawl(self, url=None):
        """
        :type url: str
        :return:
        """

        if not url:
            url = self.site_url + self.crawler_entry_point

        if self.crawler_current_videos >= self.crawler_max_videos:
            self.logger.info('Max videos number reached, end.')
            return

        prev_already_existing_videos_count = self.already_existing_videos_count

        # 1: download videos page
        [_, tree] = await self._download_videos_page(url)

        # 2: find videos from previously downloaded page
        self.logger.info('Finding videos metadata from {}...'.format(url))
        time_start = time.time()
        videos = await self._find_videos_from_videos_page(tree)
        time_end = time.time()

        self.logger.info(
            'Found videos metadata for {} videos in {:.3f} seconds.'.format(
                len(videos),
                time_end - time_start)
        )

        # 3: check if there is videos
        if not videos:
            raise ValueError('No videos found')

        # 4: find next page url from previously downloaded page
        next_page = find_next_page(tree, self.next_page_selector)

        self.logger.info('-' * 60)

        if self.already_existing_videos_count == prev_already_existing_videos_count:
            self.logger.info('0 videos were created from last crawl, now exiting...')
            return

        url = self.site_url + next_page
        await self.crawl(url)

    async def crawl_convert_video_duration_to_seconds(self, duration):
        """
        :type duration: str
        :rtype: int
        """

        raise NotImplementedError

    def _hydrate_logger(self):
        self.logger = logging.LoggerAdapter(logging.getLogger('pr0n_crawler'), {
            'site_name': self.site_name,
            'videos_current_number': self.crawler_current_videos,
            'videos_max_number': self.crawler_max_videos,
        })

    async def _download_videos_page(self, url):
        """
        :type url: str
        :rtype: list[str, lxml.html.Element]
        """

        self.logger.info('Downloading {}...'.format(url))

        time_start = time.time()
        async with aiohttp.ClientSession() as session:
            async with session.get(url) as response:
                if response.status == 404:
                    self.logger.error('Can not download {}, got 404.'.format(url))
                    exit(1)
                else:
                    self.logger.info('Downloaded in {:.3f} seconds.'.format(time.time() - time_start))
                    content = await response.text()
                    tree = html.fromstring(content)
                    return [content, tree]

    async def _find_videos_from_videos_page(self, tree):
        """
        :type tree: lxml.html.Element
        :rtype: list[Video]
        """

        videos_metadata = self._fetch_videos_page_and_find_metadata(tree)
        videos = self._get_or_create_videos_from_metadata(videos_metadata)
        await self._find_more_videos_metadata(videos)

        return videos

    def _fetch_videos_page_and_find_metadata(self, tree):
        """
        :param tree: lxml.html.Element
        :return: list of tuples following (title, url, thumbnail_url, durations) format
        :rtype: list[(str, str, str, int)]
        """

        titles = find_videos_title(tree, self.video_title_selector)
        urls = find_videos_url(tree, self.video_url_selector)
        thumbnail_urls = find_videos_thumbnail_url(tree, self.video_thumbnail_url_selector)
        durations = map(
            self.crawl_convert_video_duration_to_seconds,
            find_videos_duration(tree, self.video_duration_selector)
        )

        return list(zip(titles, urls, thumbnail_urls, durations))

    def _get_or_create_videos_from_metadata(self, videos_metadata):
        """
        :type videos_metadata: list[(str, str, str, int)]
        :rtype: list[Video]
        """
        videos = []

        for m in videos_metadata:
            video, created = Video.get_or_create(
                title=m[0].strip(),
                url=m[1].strip(),
                thumbnail_url=m[2].strip(),
                duration=m[3],
                site=self.site
            )
            videos.append(video)

            if created:
                self.already_existing_videos_count += 1

        return videos

    async def _find_more_videos_metadata(self, videos):
        """
        :param videos: list[Video]
        """

        tasks = []

        for video in videos:
            tasks.append(self._fetch_video_page_and_find_metadata(video))

        await asyncio.gather(*tasks)

    @retry(
        stop=stop_after_attempt(20), wait=wait_random(8, 512),
        before=before_log(logging.getLogger(), logging.WARN)
    )
    async def _fetch_video_page_and_find_metadata(self, video):
        """
        :type video: Video
        """

        url = self.site_url + video.url
        self.logger.info('Downloading {}...'.format(url))

        async with aiohttp.ClientSession() as session:
            async with session.get(url) as response:
                if response.status == 404:
                    self.logger.warning('Can not download {}'.format(url))
                    return

                content = await response.text()
                self.logger.info('Downloaded {}.'.format(url))
                tree = html.fromstring(content)

                details = find_video_details(tree, dict(
                    video_details_tags=self.video_details_tags_selector
                ))

                save_video_details(video, details)

                self.logger.info('Got details from {}'.format(url))
                self.crawler_current_videos += 1
                self._hydrate_logger()

    @property
    def video_title_selector(self):
        """
        :rtype: str
        """
        return self.crawler_selectors.get('video').get('title')

    @property
    def video_duration_selector(self):
        """
        :rtype: str
        """
        return self.crawler_selectors.get('video').get('duration')

    @property
    def video_url_selector(self):
        """
        :rtype: str
        """
        return self.crawler_selectors.get('video').get('url')

    @property
    def video_thumbnail_url_selector(self):
        """
        :rtype: str
        """
        return self.crawler_selectors.get('video').get('thumbnail_url')

    @property
    def video_details_tags_selector(self):
        """
        :rtype: str
        """
        return self.crawler_selectors.get('video_details').get('tags')

    @property
    def next_page_selector(self):
        """
        :rtype: str
        """
        return self.crawler_selectors.get('next_page')


def find_videos_title(tree, video_title_selector):
    """
    :type tree: lxml.html.Element
    :type video_title_selector: str
    :rtype: list[str]
    """

    return tree.xpath(video_title_selector)


def find_videos_duration(tree, video_duration_selector):
    """
    :type tree: lxml.html.Element
    :type video_duration_selector: str
    :rtype: list[str]
    """

    return tree.xpath(video_duration_selector)


def find_videos_url(tree, video_url_selector):
    """
    :type tree: lxml.html.Element
    :type video_url_selector: str
    :rtype: list[str]
    """

    return tree.xpath(video_url_selector)


def find_videos_thumbnail_url(tree, video_thumbnail_url_selector):
    """
    :type tree: lxml.html.Element
    :type video_thumbnail_url_selector: str
    :rtype: list[str]
    """

    return tree.xpath(video_thumbnail_url_selector)


def find_video_details(tree, selectors):
    """
    :type tree: lxml.html.Element
    :type selectors: dict[str, str]
    :rtype: dict[str, list]
    """

    return dict(
        tags=tree.xpath(selectors.get('video_details_tags'))
    )


def find_next_page(tree, next_page_selector):
    """
    :type tree: lxml.html.Element
    :type next_page_selector: str
    :rtype: str
    """

    return tree.xpath(next_page_selector)[0]


def save_video_details(video, details):
    """
    :type video: Video
    :type details: dict
    """

    for found_tag in details.get('tags'):
        slug = parameterize(found_tag.strip())
        tag = humanize(slug)

        if not slug:
            continue

        tag, created = Tag.get_or_create(
            tag=tag,
            slug=slug
        )

        VideoTag.get_or_create(video=video, tag=tag)
