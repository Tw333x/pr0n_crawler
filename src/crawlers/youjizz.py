from src import helpers
from src.crawlermixin import CrawlerMixin


class YoujizzCrawler(CrawlerMixin):
    site_name = 'Youjizz'
    site_url = 'https://www.youjizz.com'
    site_favicon_url = site_url + '/favicon.ico'

    crawler_entry_point = '/newest-clips/300.html'
    crawler_selectors = dict(
        prev_page='(//*[@id="pagination"]/span/preceding-sibling::a/@href)[last()]',

        video=dict(
            title='//*[contains(@class, "video-title")]/a/text()',
            duration='//*[contains(@class, "time")]/text()',
            url='//*[contains(@class, "video-title")]/a/@href',
            thumbnail_url='//a[@class="frame"]/img[contains(@class, "lazy")]/@data-original',
        ),

        video_details=dict(
            tags='//*[contains(@class, "tag-links desktop-only")]/ul/li/a/text()',
        )
    )

    def crawl_convert_video_duration_to_seconds(self, duration):
        if not duration or duration == 'N/A' or duration == 'None':
            return 0

        return helpers.hms_to_s(duration)
