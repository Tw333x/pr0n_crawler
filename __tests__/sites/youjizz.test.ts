import { readFile } from 'async-file';
import { Extractor } from '../../src/Extractor';
import { YouJizzSite } from '../../src/sites/youjizz';

const fixtures = `${__dirname}/fixtures`;

describe('Site - YouJizz', () => {
  const youjizz = new YouJizzSite();

  describe('entry point', () => {
    let extractor;

    beforeEach(async (done) => {
      const content = await readFile(`${fixtures}/youjizz-entry-point.html`, 'utf8');
      extractor = new Extractor(youjizz, content);

      done();
    });

    it('it should extracts previousPage', () => {
      expect(extractor.extractPreviousPage()).toMatchSnapshot();
    });

    it('it should extracts videosUrl', () => {
      expect(extractor.extractVideosUrl()).toMatchSnapshot();
    });

    it('it should extracts videosThumbnailUrl', () => {
      expect(extractor.extractVideosThumbnailUrl()).toMatchSnapshot();
    });
  });
});
