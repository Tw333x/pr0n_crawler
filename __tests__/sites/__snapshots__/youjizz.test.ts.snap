// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site - YouJizz entry point it should extracts previousPage 1`] = `"https://www.youjizz.com/newest-clips/99.html"`;

exports[`Site - YouJizz entry point it should extracts videosThumbnailUrl 1`] = `
Array [
  "https://cdne-pics.youjizz.com/f/4/7/f47dc3f9ee5f0930264933945c8298541496219703-960-540-2099-h264.mp4-3.jpg",
  "https://cdne-pics.youjizz.com/a/0/8/a084a3bdc08964db9a42b5b30c8620461513059926-1920-1080-9876-h264.mp4-4.jpg",
  "https://cdne-pics.youjizz.com/6/7/8/678e5b2efab5ceb531a2d807e5180d701475248211-1280-720-1853-h264.mp4-3.jpg",
  "https://cdne-pics.youjizz.com/e/a/e/eaeb1f4ab43bdc7c3699e2e4dfd950f51512972016-1920-1080-9984-h264.mp4-8.jpg",
  "https://cdne-pics.youjizz.com/0/8/e/08e1ff2c7c01426cf95e860b5793f4161510985008-1920-1080-8331-h264.mp4-7.jpg",
  "https://cdne-pics.youjizz.com/3/2/4/32434d975828b58f0623c7a76a5be3a11497168333-1280-720-2050-h264.mp4-2.jpg",
  "https://cdne-pics.youjizz.com/4/5/5/4558ebcdb8d40efde53385888b7f1ebf1510812055-1920-1080-8563-h264.mp4-7.jpg",
  "https://cdne-pics.youjizz.com/f/d/1/fd14b627b3a54ba3fe51bcafc0f2267e1484338565-720-404-1108-h264.mp4-6.jpg",
  "https://cdne-pics.youjizz.com/4/8/8/488d220a2b55823fa8112ac2f3b57ed61447747805-480-270-401-h264.flv-5.jpg",
  "https://cdne-pics.youjizz.com/9/a/4/9a41b2275275944417f28724e3607ed91496269811-1280-720-1448-h264.mp4-5.jpg",
  "https://cdne-pics.youjizz.com/c/9/6/c968f1c71c11559341a4a8b74b4591db1493595307-640-360-598-h264.mp4-5.jpg",
  "https://cdne-pics.youjizz.com/d/0/4/d0483d6dff746b1aa4cee75c656ade5d1478900706-1280-720-2048-h264.mp4-6.jpg",
  "https://cdne-pics.youjizz.com/5/1/e/51ea5623c0f67474e2985b1d7d8a00bd1512194407-1920-1080-8390-h264.mp4-3.jpg",
  "https://cdne-pics.youjizz.com/2/b/0/2b0612b41a3089dc7b378d8d0b1409761472776203-334-240-512-h264.mp4-5.jpg",
  "https://cdne-pics.youjizz.com/7/9/e/79e536cb667526f4952c862709b157e91490593526-1280-720-2046-h264.mp4-3.jpg",
  "https://cdne-pics.youjizz.com/1/5/e/15e34e99efc20bb521c151bfb5fc05c21493578831-1280-720-4002-h264.mp4-7.jpg",
  "https://cdne-pics.youjizz.com/e/d/2/ed26964e32771fa7e56a1f2c2644750e1501497306-512-336-193-h264.mp4-7.jpg",
  "https://cdne-pics.youjizz.com/d/1/9/d19c50dce196c8df5240f751cd93b6641478983529-1280-720-1539-h264.mp4-4.jpg",
  "https://cdne-pics.youjizz.com/e/6/4/e64dbe78ffb03373759d127c0a293ff61480146369-1280-720-1413-h264.mp4-5.jpg",
  "https://cdne-pics.youjizz.com/b/b/d/bbd8d61579de567f4100a8b57a6b98471495051521-1280-720-2500-h264.mp4-6.jpg",
  "https://cdne-pics.youjizz.com/6/3/4/634039586f376665402dc5b7b7c6e0181473792020-320-240-510-h264.mp4-5.jpg",
  "https://cdne-pics.youjizz.com/7/f/2/7f2ac10f029369347e287942122705ca1497582004-360-240-512-h264.mp4-8.jpg",
  "https://cdne-pics.youjizz.com/c/5/0/c505d43933cf28fd93228b96193745841487537180-480-320-466-h264.mp4-4.jpg",
  "https://cdne-pics.youjizz.com/2/9/d/29d437ae9670ebdd734b261a42b11bfc1479379537-1280-720-1855-h264.mp4-5.jpg",
  "https://cdne-pics.youjizz.com/f/4/7/f47dc3f9ee5f0930264933945c8298541496219703-960-540-2099-h264.mp4-3.jpg",
  "https://cdne-pics.youjizz.com/a/0/8/a084a3bdc08964db9a42b5b30c8620461513059926-1920-1080-9876-h264.mp4-4.jpg",
  "https://cdne-pics.youjizz.com/6/7/8/678e5b2efab5ceb531a2d807e5180d701475248211-1280-720-1853-h264.mp4-3.jpg",
  "https://cdne-pics.youjizz.com/e/a/e/eaeb1f4ab43bdc7c3699e2e4dfd950f51512972016-1920-1080-9984-h264.mp4-8.jpg",
  "https://cdne-pics.youjizz.com/0/8/e/08e1ff2c7c01426cf95e860b5793f4161510985008-1920-1080-8331-h264.mp4-7.jpg",
  "https://cdne-pics.youjizz.com/3/2/4/32434d975828b58f0623c7a76a5be3a11497168333-1280-720-2050-h264.mp4-2.jpg",
  "https://cdne-pics.youjizz.com/4/5/5/4558ebcdb8d40efde53385888b7f1ebf1510812055-1920-1080-8563-h264.mp4-7.jpg",
  "https://cdne-pics.youjizz.com/f/d/1/fd14b627b3a54ba3fe51bcafc0f2267e1484338565-720-404-1108-h264.mp4-6.jpg",
  "https://cdne-pics.youjizz.com/4/8/8/488d220a2b55823fa8112ac2f3b57ed61447747805-480-270-401-h264.flv-5.jpg",
  "https://cdne-pics.youjizz.com/9/a/4/9a41b2275275944417f28724e3607ed91496269811-1280-720-1448-h264.mp4-5.jpg",
  "https://cdne-pics.youjizz.com/c/9/6/c968f1c71c11559341a4a8b74b4591db1493595307-640-360-598-h264.mp4-5.jpg",
  "https://cdne-pics.youjizz.com/d/0/4/d0483d6dff746b1aa4cee75c656ade5d1478900706-1280-720-2048-h264.mp4-6.jpg",
  "https://cdne-pics.youjizz.com/5/1/e/51ea5623c0f67474e2985b1d7d8a00bd1512194407-1920-1080-8390-h264.mp4-3.jpg",
  "https://cdne-pics.youjizz.com/2/b/0/2b0612b41a3089dc7b378d8d0b1409761472776203-334-240-512-h264.mp4-5.jpg",
  "https://cdne-pics.youjizz.com/7/9/e/79e536cb667526f4952c862709b157e91490593526-1280-720-2046-h264.mp4-3.jpg",
  "https://cdne-pics.youjizz.com/1/5/e/15e34e99efc20bb521c151bfb5fc05c21493578831-1280-720-4002-h264.mp4-7.jpg",
  "https://cdne-pics.youjizz.com/e/d/2/ed26964e32771fa7e56a1f2c2644750e1501497306-512-336-193-h264.mp4-7.jpg",
  "https://cdne-pics.youjizz.com/d/1/9/d19c50dce196c8df5240f751cd93b6641478983529-1280-720-1539-h264.mp4-4.jpg",
  "https://cdne-pics.youjizz.com/e/6/4/e64dbe78ffb03373759d127c0a293ff61480146369-1280-720-1413-h264.mp4-5.jpg",
  "https://cdne-pics.youjizz.com/b/b/d/bbd8d61579de567f4100a8b57a6b98471495051521-1280-720-2500-h264.mp4-6.jpg",
  "https://cdne-pics.youjizz.com/6/3/4/634039586f376665402dc5b7b7c6e0181473792020-320-240-510-h264.mp4-5.jpg",
  "https://cdne-pics.youjizz.com/7/f/2/7f2ac10f029369347e287942122705ca1497582004-360-240-512-h264.mp4-8.jpg",
  "https://cdne-pics.youjizz.com/c/5/0/c505d43933cf28fd93228b96193745841487537180-480-320-466-h264.mp4-4.jpg",
  "https://cdne-pics.youjizz.com/2/9/d/29d437ae9670ebdd734b261a42b11bfc1479379537-1280-720-1855-h264.mp4-5.jpg",
]
`;

exports[`Site - YouJizz entry point it should extracts videosUrl 1`] = `
Array [
  "https://www.youjizz.com/videos/having-fun-with-boyfriends-friend-41212501.html",
  "https://www.youjizz.com/videos/i-am-going-to-fuck-your-wife-p3-44980291.html",
  "https://www.youjizz.com/videos/hottest-amateur-pov-sex-scene-33226821.html",
  "https://www.youjizz.com/videos/petite-asian-squirter-44957911.html",
  "https://www.youjizz.com/videos/the-joys-of-a-long-hot-shower-44591631.html",
  "https://www.youjizz.com/videos/hardcore-threesome-in-the-house-41460901.html",
  "https://www.youjizz.com/videos/nail-her-p1-44554291.html",
  "https://www.youjizz.com/videos/how-big-is-it---stassi-sinclair-36508821.html",
  "https://www.youjizz.com/videos/first-fuck-28912211.html",
  "https://www.youjizz.com/videos/perfect-body-teenie-rubbing-her-pussy-on-old-guy-41228271.html",
  "https://www.youjizz.com/videos/the-japanese-kind-of-shower-massage-40422391.html",
  "https://www.youjizz.com/videos/two-teen-stepsister-get-first-anal-fuck-by-big-dick-step-bro-34328731.html",
  "https://www.youjizz.com/videos/zz-pizza-party-part-1-p2-44808671.html",
  "https://www.youjizz.com/videos/old-french-mature-offers-her-ass-32761981.html",
  "https://www.youjizz.com/videos/return-of-the-harem-girl-39242121.html",
  "https://www.youjizz.com/videos/blonde-milf-only-takes-bbc-40422451.html",
  "https://www.youjizz.com/videos/girlfriend-gets-analized-and-facialized-42458801.html",
  "https://www.youjizz.com/videos/two-sexy-blondes-develop-a-taste-for-pussy-34358221.html",
  "https://www.youjizz.com/videos/stunning-cowgirl-ride-on-cock-34745021.html",
  "https://www.youjizz.com/videos/jessica-red-masturbating-40834471.html",
  "https://www.youjizz.com/videos/hot-bbw-in-red-32941081.html",
  "https://www.youjizz.com/videos/girl-on-girl-687-41552931.html",
  "https://www.youjizz.com/videos/brunette-babe-lenka-loyosice-37840011.html",
  "https://www.youjizz.com/videos/seline-benazir-gets-horny-and-fucks-herself-with-a-gigantic-34514031.html",
  "https://www.youjizz.com/videos/having-fun-with-boyfriends-friend-41212501.html",
  "https://www.youjizz.com/videos/i-am-going-to-fuck-your-wife-p3-44980291.html",
  "https://www.youjizz.com/videos/hottest-amateur-pov-sex-scene-33226821.html",
  "https://www.youjizz.com/videos/petite-asian-squirter-44957911.html",
  "https://www.youjizz.com/videos/the-joys-of-a-long-hot-shower-44591631.html",
  "https://www.youjizz.com/videos/hardcore-threesome-in-the-house-41460901.html",
  "https://www.youjizz.com/videos/nail-her-p1-44554291.html",
  "https://www.youjizz.com/videos/how-big-is-it---stassi-sinclair-36508821.html",
  "https://www.youjizz.com/videos/first-fuck-28912211.html",
  "https://www.youjizz.com/videos/perfect-body-teenie-rubbing-her-pussy-on-old-guy-41228271.html",
  "https://www.youjizz.com/videos/the-japanese-kind-of-shower-massage-40422391.html",
  "https://www.youjizz.com/videos/two-teen-stepsister-get-first-anal-fuck-by-big-dick-step-bro-34328731.html",
  "https://www.youjizz.com/videos/zz-pizza-party-part-1-p2-44808671.html",
  "https://www.youjizz.com/videos/old-french-mature-offers-her-ass-32761981.html",
  "https://www.youjizz.com/videos/return-of-the-harem-girl-39242121.html",
  "https://www.youjizz.com/videos/blonde-milf-only-takes-bbc-40422451.html",
  "https://www.youjizz.com/videos/girlfriend-gets-analized-and-facialized-42458801.html",
  "https://www.youjizz.com/videos/two-sexy-blondes-develop-a-taste-for-pussy-34358221.html",
  "https://www.youjizz.com/videos/stunning-cowgirl-ride-on-cock-34745021.html",
  "https://www.youjizz.com/videos/jessica-red-masturbating-40834471.html",
  "https://www.youjizz.com/videos/hot-bbw-in-red-32941081.html",
  "https://www.youjizz.com/videos/girl-on-girl-687-41552931.html",
  "https://www.youjizz.com/videos/brunette-babe-lenka-loyosice-37840011.html",
  "https://www.youjizz.com/videos/seline-benazir-gets-horny-and-fucks-herself-with-a-gigantic-34514031.html",
]
`;
